feat: Fastembed - add FastembedRanker

paulmartrencharpro opened the pull request:
### Proposed Changes:
New feature: integrate with the new ranker from fastembed. (https://qdrant.tech/articles/cross-encoder-integration-gsoc/, https://github.com/qdrant/fastembed/tree/main?tab=readme-ov-file#-rerankers)

I based my code on CohereRanker and Fastembed existing code

### How did you test it?

I wrote all the tests in test_fastembed_ranker.py, based on the other tests for fastembed.

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

anakin87 commented on integrations/fastembed/src/haystack_integrations/components/rankers/fastembed/ranker.py:
```diff
@@ -0,0 +1,194 @@
+    docs = [Document(content="Paris"), Document(content="Berlin")]
+    query = "What is the capital of germany?"
+    output = ranker.run(query=query, documents=docs)
+    docs = output["documents"]
```
```suggestion
   print(output["documents"][0].content)
   
   # Berlin
```

I think adding an example of the output can help.

anakin87 commented on integrations/fastembed/src/haystack_integrations/components/rankers/fastembed/ranker.py:
```diff
@@ -0,0 +1,194 @@
+        :param meta_data_separator: Separator used to concatenate the meta fields
+            to the Document content.
+        """
+        self.model_name = model_name
```
```suggestion
        if top_k <= 0:
            raise ValueError(f"top_k must be > 0, but got {top_k}")
            
        self.model_name = model_name
```

let's add this check

anakin87 commented on integrations/fastembed/src/haystack_integrations/components/rankers/fastembed/ranker.py:
```diff
@@ -0,0 +1,194 @@
+    @component.output_types(documents=List[Document])
+    def run(self, query: str, documents: List[Document], top_k: Optional[int] = None):
+        """
+        Use the Fastembed Reranker to re-rank the list of documents based on the query using fastembed
+
+        :param query:
+            Query string.
+        :param documents:
+            List of Documents.
+        :param top_k:
+            The maximum number of Documents you want the Ranker to return.
+        :returns:
+            A dictionary with the following keys:
+            - `documents`: List of Documents most similar to the given query in descending order of similarity.
```
```suggestion
        Returns a list of documents ranked by their similarity to the given query, using FastEmbed.

        :param query:
            The input query to compare the documents to.
        :param documents:
            A list of documents to be ranked.
        :param top_k:
            The maximum number of documents to return.

        :returns:
            A dictionary with the following keys:
            - `documents`: A list of documents closest to the query, sorted from most similar to least similar.
```

anakin87 commented on integrations/fastembed/src/haystack_integrations/components/rankers/fastembed/ranker.py:
```diff
@@ -0,0 +1,194 @@
+            msg = "No query provided"
+            raise ValueError(msg)
+
+        if documents.count == 0:
+            msg = "No documents provided"
+            raise ValueError(msg)
```
```suggestion
        if not documents:
            return {"documents": []}
```

we usually don't raise an error in this case

anakin87 commented on integrations/fastembed/src/haystack_integrations/components/rankers/fastembed/ranker.py:
```diff
@@ -0,0 +1,194 @@
+        """
+        Creates an instance of the 'FastembedRanker'.
+
+        :param model_name: Fastembed model name. Check the list of supported models in the [Fastembed documentation](https://docs.Fastembed.com/docs/models).
```
wrong URL

paulmartrencharpro replied:
Indeed. I changed it
Note: the models are not there yet. I made a PR for this on the fastembed part (https://github.com/qdrant/fastembed/pull/393). It's merged, but I think it needs a release to be pushed on the website

anakin87 reviewed (changes requested):
Hey! Thank you for this PR.

I've had a first look and it seems good.

I have an idea: in general, we have an embedding backend for local Embedders because we expect users to use both Text Embedder and Doc Embedder based on the same model; this saves resources (the model is loaded only once).

For rankers, we don't have this problem.
So I would ask you to **remove the backend and do everything in the component** (Cohere can be taken as example).

Let me know and thanks again!

anakin87 commented on integrations/fastembed/tests/test_fastembed_ranker.py:
```diff
@@ -0,0 +1,235 @@
+        assert ranker.meta_fields_to_embed == ["test_field"]
+        assert ranker.meta_data_separator == " | "
+
+    def test_embed_incorrect_input_format(self):
+        """
+        Test for checking incorrect input format when creating embedding.
+        """
```
```suggestion
    def test_run_incorrect_input_format(self):
        """
        Test for checking incorrect input format.
        """
```

anakin87 commented on integrations/fastembed/tests/test_fastembed_ranker.py:
```diff
@@ -0,0 +1,235 @@
+
+    def test_embed_metadata(self):
+        """
+        Test for checking output dimensions and embedding dimensions for documents
+        with a custom instruction and metadata.
```
```suggestion
        Tests the embedding of metadata fields in document content for ranking.
```

anakin87 commented on integrations/fastembed/tests/test_fastembed_ranker.py:
```diff
@@ -0,0 +1,235 @@
+
+        assert isinstance(first_document, Document)
+        assert first_document.content == "fastembed is supported by and maintained by Qdrant."
+        assert first_document.score == 5.472434997558594
```
Let's only check that first document score > second document score.
Otherwise, this test can become flaky due to temporary circumstances and external changes.

paulmartrencharpro replied:
Done

paulmartrencharpro commented:
> In general, this PR is almost ready to be merged.
> 
> TODO:
> 
>     * modify this line in README to mention ranker: https://github.com/deepset-ai/haystack-core-integrations/blob/c3437763e7c80e11e9ed8678f1b235353a316781/README.md?plain=1#L37
> 
>     * add more tests if you can/want (uncovered lines: 66-67, 109, 115->exit, 162-163, 166, 170-171, 174-175)
> 
> 
> Then, if you also want to open a PR on haystack-integrations to update this page ( https://github.com/deepset-ai/haystack-integrations/blob/main/integrations/fastembed.md), it would be appreciated - otherwise, I'll take care of it.

I updated the Readme and added tests.
The only test I cannot write is the one that check that we don't create 2 TextCrossEncoder when we call warm_up multiple times. I don't understand the patch/mockup code to replicate it. I wrote this but it's not working:

    @patch('fastembed.rerank.cross_encoder.TextCrossEncoder')
    def test_warmup_twice(self, mock_encoder):
        """
        Test for checking that warm up only creates one TextCrossEncoder
        """
        ranker = FastembedRanker(model_name="Xenova/ms-marco-MiniLM-L-12-v2")
        ranker.warm_up()
        ranker.warm_up()

        # Assert that TextCrossEncoder is called only once
        mock_encoder.assert_called_once()
        

I'll make the PR on the other repo

anakin87 reviewed (approved):
Everything looks good now!

Thanks for the good work. :heart_decoration: 

(I'll track the related stuff in https://github.com/deepset-ai/haystack-core-integrations/issues/1188)
